### ISSUE TITLE
Why not use KL-divergence or cross-entropy as the objective function?

### ISSUE THREAD

**Issue by Lecanyu** [xu-ji/IIC#30]
Hi, I have read your paper and I'm interested in your approach. In my opinion, the invariant information shares  similar idea with consistency training in semi-supervised learning [1]. 

In [1], they use KL-divergence to force consistent prediction between data and augmented ones. I think KL-divergence may be also proper in your implementation without degenracy. 

Have you tried some experiments on KL-divergence or cross-entropy as objective function? Could you explain the differences among them?

[1] Unsupervised Data Augmentation for Consistency Training, Xie et al.

**Comment by xu-ji** [comment]
Thank you for the insightful question. 

Since cross-entropy(P1, P2)= entropy(P1) + KL(P1 || P2), minimising it would actually be minimising entropy of predictions, which would not help us avoid degeneracy, as we want to maximise entropy of predictions. For example, if all images were mapped to [1, 0, 0] prediction, cross-entropy is still minimised.

KL(P1 || P2) = - entropy(P1) + cross-entropy(P1, P2), so minimising it would protect against degeneracy, but it is also not symmetric. We want symmetry because the order of the paired variables should not matter in our setup. 

Mutual information involves maximising entropy and is also symmetric.

**Comment by Lecanyu** [comment]
Hi, thanks for your clarification.

Yes, cross-entropy objective will lead to degeneracy. 

However, the symmetric problem in KL-divergence may be negligible since the data pair are randomly generated during the training procedure (i.e. both pairs [image1, image2] and [image2, image1] will be accessed, the model will be trained insensitive to the image order within a pair).

It would be nice if you can conduct some experiments on KL-divergence objective I think.

Besides, I'm still a little bit confused about the intuition of mutual information. Mutual information I(P1 ; P2) can be maximized when both P1 and P2 have uniform distribution on all classes. But if they are uniform distribution, the model has no discriminative ability, right?

**Comment by xu-ji** [comment]
Actually in our implementation image1 is always the original image and image2 is always the transformed one, so if KL was used the gradients coming from the two sides would be consistently non-symmetric. It's nice that with mutual information you do not need to care about this experimental detail.

It's all about order actually. KL uses cross entropy non-symmetrically, as KL(X || Y) = - H(X) + H(X, Y) **!=** - H(Y) + H(Y, X). Whereas MI uses conditional entropy symmetrically, as MI(X, Y) = H(X) - H(X | Y) = H(Y) - H(Y | X) (so maxing MI minimises both conditional entropies H(X | Y) and H(Y | X)). Therefore in KL, you need to assume one signal (e.g. X) is more trustworthy than the other, but in MI you don't. This means with KL, X and Y are not symmetrically encouraged to be one-hot, whereas with MI it is symmetric encouragement.

As for the reason why MI encourages the distributions to be one-hot (i.e. fully discriminative model), see section 11 of the [supplementary material](https://github.com/xu-ji/IIC/blob/master/paper/suppmat_camera_ready.pdf) for a detailed example. Essentially, yes MI encourages uniform predictions _across the dataset_ (i.e. marginalised across samples), but if individual predictions (conditioned on a single sample) were uniform, conditional entropy is not minimised. So for dataset of 4 images and 2 clusters, [1, 0], [0, 1], [0, 1], [1, 0] is optimal but [0.5, 0.5], [0.5, 0.5], [0.5, 0.5], [0.5, 0.5] is not.

It is worth adding here that maxing MI means finding the best balance of entropy against conditional entropy, for a given dataset and architecture. If, for example, reducing prediction entropy a little means you could get a larger reduction in conditional entropy, the optimisation will take that step. This means maxing MI is not simply a pure optimisation for equal cluster assignment.

We certainly could try KL in future work. But it would really be out of just curiosity, since MI already captures exactly what we want from the model and KL doesn't, with respect to signal trust.

**Comment by jizongFox** [comment]
In the clustering setting, Kl is worse than mi, based on my you experimental results
